File: pset6/credit/credit.py
from cs50 import get_int

cc = get_int("Number: ")
sum = 0

# specify first number index
digit_index = len(str(cc)) - 2

# First numbers loop
while digit_index >= 0:
    digit = int(str(cc)[digit_index]) * 2

    if digit > 9:
        sum += int(str(digit)[0]) + int(str(digit)[1])
    else:
        sum += digit

    digit_index -= 2

# redefine digit index for second numbers iteration
digit_index = len(str(cc)) - 1

# Second numbers loop
while digit_index >= 0:
    sum += int(str(cc)[digit_index])
    digit_index -= 2

# variables for vendor check
visa = int(str(cc)[0]) == 4
amex = int(str(cc)[:2]) == 34 or int(str(cc)[:2]) == 37
mc = int(str(cc)[:2]) in range(51, 56)

# Check if card is valid
if sum % 10 != 0:
    print("INVALID")
else:
    # vendor check
    if amex:
        print("AMEX")
    elif visa:
        print("VISA")
    elif mc:
        print("MASTERCARD")

File: pset1/credit/credit.c
#include <stdio.h>
#include <cs50.h>
#include <math.h>

// Counts total digit count
int digit_count(long long n)
{
    if (n == 0)
    {
        return 0;
    }
    return 1 + digit_count(n / 10);
}

int main(void)
{
    long long cc_number = get_long_long("Enter CC #: ");
    int total_digits = digit_count(cc_number);
    int sum = 0;

    // Looping  throught first set of numbers, doubling them and adding to sum
    long first_division = 10;
    for (int i = 0; i < total_digits / 2; i++)
    {
        int current_number_doubled = ((cc_number / first_division) % 10) * 2;

        // If doubled number ends up being two digit number run another loop
        // which adds digits one by one
        if (current_number_doubled > 9)
        {
            int local_multiplier = 1;
            for (int j = 0; j < 2; j++)
            {
                sum += (current_number_doubled / local_multiplier) % 10;
                local_multiplier = 10;
            }
        }
        else
        {
            sum += current_number_doubled;
        }
        first_division *= 100;
    }

    // Run second loop which adds non multiplied digits to sum
    long second_division = 1;
    for (int i = 0; i < total_digits; i += 2)
    {
        int current_number = (cc_number / second_division) % 10;
        sum += current_number;
        second_division *= 100;
    }

    // Check for a first digit to determine whether it is a valid
    // American Express, MasterCard, or Visa card number
    int first_number = (long long)(cc_number / pow(10, total_digits - 1));
    int two_first_numbers = (long long)(cc_number / pow(10, total_digits - 2));

    // Validation check
    if (sum % 10 != 0)
    {
        printf("INVALID\n");
    }
    else
    {
        // Vendor check
        if (two_first_numbers == 34 || two_first_numbers == 37)
        {
            printf("AMEX\n");
        }
        else if (two_first_numbers == 51 || two_first_numbers == 52 || two_first_numbers == 53 || two_first_numbers == 54
                 || two_first_numbers == 55)
        {
            printf("MASTERCARD\n");
        }
        else if (first_number == 4 && total_digits > 12)
        {
            printf("VISA\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
}